### ISSUE TITLE
cost effective backpropagating through an iterative process

### ISSUE THREAD

**Issue by OrGreenberg** [dvirsamuel/SeedSelect#3]
Hi, Thank you for the great paper!

During the optimization step of SeedSelect, you are actually backpropagating through T iterations of the very same Unet (using either DDIM or DDPM sampling). Do you unroll the iterative process in order to prevent over-writing the derivatives of iteration 't' by the ones of 't-1'? Considering that all the derivatives are needed for the backward pass of SeedSelect's otimization step, how do you deal with the memory cost of memorizing all those derivatives (at least 50 times of SD)?

Did you run into memory issues during the process? can you please share your memory usage at the pick right before the optimization step?

Once again, thank you for both of these papers.

**Comment by dvirsamuel** [comment]
Hi Or,
Thank you for your interest in our work.

Yes, we backpropagate through all steps. Note that we actually used 20 denoising steps to generate an image. In the paper we mentioned that by using Euler Scheduler, it is possible to generate an image with fewer than 10 denoising steps, resulting in a significant reduction in GPU memory usage.

We further used gradient checkpointing (see [here](https://github.com/dvirsamuel/SeedSelect/blob/main/SeedOptimizationStableDiffusion.py#L34)) which enables us to use less than 35GB VRAM during optimization with 20 denoising steps.

Dvir
